### DISCUSSION TITLE
Perhaps loosen strictness on LocationBase equality check grids?

### DISCUSSION THREAD

**Discussion by drewj-tp** [terrapower/armi#1942] [Ideas]
I stumbled on this weird quirk working on assembly rotation: If I deepcopy a block with a grid and grab the same component from the original and copied block, their locations will not be equal. In testing, I got weird errors like
```
<IndexLocation @ (0,0,0)> != <IndexLocation @ (0,0,0)>
```
which is confusing at first. These are both the center location on a hexagonal grid of the same pitch. Their locations should compare equal?

Turns out, two locations must have the exact same grid object attached, not an equivalent grid

https://github.com/terrapower/armi/blob/c6f46e5d04dc265e7a8e470fdd1977c9a4e16e3e/armi/reactor/grids/locations.py#L100-L109

Would it make sense to have the check on grids be `self.grids == other.grids`? We don't currently provide `Grid.__eq__` so that'd be a need

**Comment by john-science** [comment]
My take is that there was no `Grid.__eq__()` method or `SpatialGrid.__eq__()` method, so someone got lazy. We'd have to look at how to compare grids, but if that's doable, do it. Sure.

**Reply by drewj-tp** [reply]
It _should_ be relatively easy? Same type and same pitch? And I guess orientation for hex grids?

**Reply by drewj-tp** [reply]
But we also have these internal `_locations` default-dict like attributes that store indices for things that may or may not be on the grid. Should those factor into the equality?

https://github.com/terrapower/armi/blob/d3cc8e5fcb1f83206e372baa657b3d51a7f6496c/armi/reactor/grids/structuredGrid.py#L263-L280

So you could have two hex grids of the same pitch and orientation, but their length, as reported by `__len__`

https://github.com/terrapower/armi/blob/d3cc8e5fcb1f83206e372baa657b3d51a7f6496c/armi/reactor/grids/structuredGrid.py#L182-L183

could differ. And that feels weird?
